lib/pbench/test/unit/server/query_apis/test_datasets_update.py

```diff
@@ -228,10 +228,134 @@ def fake_bulk(
         monkeypatch.setattr("elasticsearch.helpers.streaming_bulk", fake_bulk)
 
         response = client.post(
-            f"{server_config.rest_uri}/datasets/publish/random_md5_string1",
+            f"{server_config.rest_uri}/datasets/random_md5_string1",
             headers={"authorization": f"Bearer {pbench_token}"},
             query_string=self.PAYLOAD,
         )
 
         # Verify the failure
         capinternal("Unexpected backend error", response)
+
+    @pytest.mark.parametrize(
+        "owner,access",
+        [
+            ("drb", "public"),
+            ("drb", "private"),
+            ("test", "public"),
+            ("test", "private"),
+        ],
+    )
+    def test_query_owner_publish(
+        self,
+        access,
+        attach_dataset,
+        build_auth_header,
+        client,
+        create_drb_user,
+        create_user,
+        get_document_map,
+        monkeypatch,
+        owner,
+        server_config,
+    ):
+        """
+        Check behavior of the datasets_update API with various combinations of dataset
+        access and owner (managed by the "owner" parametrization here) and
+        authenticated user (managed by the build_auth_header fixture).
+        """
+        self.fake_elastic(monkeypatch, get_document_map, False)
+
+        is_admin = build_auth_header["header_param"] == HeaderTypes.VALID_ADMIN
+        if not HeaderTypes.is_valid(build_auth_header["header_param"]):
+            expected_status = HTTPStatus.UNAUTHORIZED
+        elif not is_admin:
+            expected_status = HTTPStatus.FORBIDDEN
+        else:
+            expected_status = HTTPStatus.OK
+
+        ds = Dataset.query(name=owner)
+        assert_id = str(create_drb_user.id) if owner == "drb" else str(create_user.id)
+        response = client.post(
+            f"{server_config.rest_uri}/datasets/{ds.resource_id}",
+            headers=build_auth_header["header"],
+            query_string={"owner": owner, "access": access},
+        )
+
+        assert response.status_code == expected_status
+        if expected_status == HTTPStatus.OK:
+            assert response.json == {"ok": 31, "failure": 0}
+            dataset = Dataset.query(name=owner)
+            assert dataset.access == access
+            assert dataset.owner_id == assert_id
+
+    def test_invalid_owner_params(
+        self,
+        client,
+        get_document_map,
+        monkeypatch,
+        pbench_admin_token,
+        server_config,
+    ):
+        """
+        Check the datasets_update API response if the "owner" attribute is invalid.
+        """
+
+        ds = Dataset.query(name="drb")
+        response = client.post(
+            f"{server_config.rest_uri}/datasets/{ds.resource_id}",
+            headers={"authorization": f"Bearer {pbench_admin_token}"},
+            query_string={"owner": str("invalid_owner")},
+        )
+
+        # Verify the report and status
+        assert response.status_code == HTTPStatus.NOT_FOUND
```

Comment:
`404`/`NOT_FOUND` seems like the wrong response code in this instance.  The _resource_ (i.e., the dataset) is there -- it's the proposed new _owner_ which is unavailable.

I don't have a good solution for this, but it strikes me that `400`/`BAD REQUEST` is probably a better fit.

(If the `404` is being generated deep within the code rather than near the top level of the `datasets_update` `POST` method, then perhaps nevermind.)

Reply:
This is wired into `convert_username` as a result of team discussions long ago in a galaxy ... well, not so far from where we are right now. The logic was that a user is a "resource" (though that was argued, and becomes somewhat tenuous once we've completed the Keycloak transition and it's no longer *our* resource...)